From patchwork Thu Jan 17 12:40:12 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Anisse Astier <aastier@freebox.fr>
X-Patchwork-Id: 10768011
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id EA5131580
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:28 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id D9E9A2F15C
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:28 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id CE0412F239; Thu, 17 Jan 2019 12:40:28 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 76A172F229
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:28 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727786AbfAQMk0 (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 17 Jan 2019 07:40:26 -0500
Received: from smtp3-g21.free.fr ([212.27.42.3]:62897 "EHLO smtp3-g21.free.fr"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1727773AbfAQMk0 (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 17 Jan 2019 07:40:26 -0500
Received: from anisse-station.iliad.local (unknown [213.36.7.13])
        by smtp3-g21.free.fr (Postfix) with ESMTPS id DAF5B13F8E4;
        Thu, 17 Jan 2019 13:40:20 +0100 (CET)
From: Anisse Astier <aastier@freebox.fr>
To: kvm@vger.kernel.org
Cc: Anisse Astier <aastier@freebox.fr>
Subject: [PATCH kvmtool 1/3] builtin-run: Fix warning when resolving path
Date: Thu, 17 Jan 2019 13:40:12 +0100
Message-Id: <20190117124014.18308-2-aastier@freebox.fr>
X-Mailer: git-send-email 2.19.1
In-Reply-To: <20190117124014.18308-1-aastier@freebox.fr>
References: <20190117124014.18308-1-aastier@freebox.fr>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

GCC 8.2 gives this warning:

builtin-run.c: In function ‘kvm_run_write_sandbox_cmd.isra.1’:
builtin-run.c:417:28: error: ‘%s’ directive output may be truncated writing up to 4095 bytes into a region of size 4091 [-Werror=format-truncation=]
   snprintf(dst, len, "/host%s", resolved_path);
                            ^~   ~~~~~~~~~~~~~

It's because it understands that len is PATH_MAX, the same as
resolved_path's size. This patch handles the case where the string is
truncated, and fixes the warning.

Signed-off-by: Anisse Astier <aastier@freebox.fr>
---
 builtin-run.c | 4 +++-
 1 file changed, 3 insertions(+), 1 deletion(-)

diff --git a/builtin-run.c b/builtin-run.c
index 443c10b..32d8aef 100644
--- a/builtin-run.c
+++ b/builtin-run.c
@@ -414,7 +414,9 @@ static void resolve_program(const char *src, char *dst, size_t len)
 		if (!realpath(src, resolved_path))
 			die("Unable to resolve program %s: %s\n", src, strerror(errno));
 
-		snprintf(dst, len, "/host%s", resolved_path);
+		if (snprintf(dst, len, "/host%s", resolved_path) >= (int)len - 1)
+			die("Pathname too long: %s -> %s\n", src, resolved_path);
+
 	} else
 		strncpy(dst, src, len);
 }

From patchwork Thu Jan 17 12:40:13 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Anisse Astier <aastier@freebox.fr>
X-Patchwork-Id: 10768013
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 66DCA6C2
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:32 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 588AE2F059
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:32 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 56BCF2F26D; Thu, 17 Jan 2019 12:40:32 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id BEA052EFB5
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:31 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727803AbfAQMka (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 17 Jan 2019 07:40:30 -0500
Received: from smtp3-g21.free.fr ([212.27.42.3]:63266 "EHLO smtp3-g21.free.fr"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1727790AbfAQMka (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 17 Jan 2019 07:40:30 -0500
Received: from anisse-station.iliad.local (unknown [213.36.7.13])
        by smtp3-g21.free.fr (Postfix) with ESMTPS id E8C5413F8E4;
        Thu, 17 Jan 2019 13:40:24 +0100 (CET)
From: Anisse Astier <aastier@freebox.fr>
To: kvm@vger.kernel.org
Cc: Anisse Astier <aastier@freebox.fr>
Subject: [PATCH kvmtool 2/3] virtio: fix warning on strncpy
Date: Thu, 17 Jan 2019 13:40:13 +0100
Message-Id: <20190117124014.18308-3-aastier@freebox.fr>
X-Mailer: git-send-email 2.19.1
In-Reply-To: <20190117124014.18308-1-aastier@freebox.fr>
References: <20190117124014.18308-1-aastier@freebox.fr>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

GCC 8.2 gives this warning:

virtio/net.c: In function ‘virtio_net__tap_init’:
virtio/net.c:336:47: error: argument to ‘sizeof’ in ‘strncpy’ call is the same expression as the source; did you mean to use the size of the destination? [-Werror=sizeof-pointer-memaccess]
   strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
                                               ^
virtio/net.c:348:47: error: argument to ‘sizeof’ in ‘strncpy’ call is the same expression as the source; did you mean to use the size of the destination? [-Werror=sizeof-pointer-memaccess]
   strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
                                               ^

Fix it by using sizeof of destination instead, even if they're the same
size in this case.

Signed-off-by: Anisse Astier <aastier@freebox.fr>
---
 virtio/net.c | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/virtio/net.c b/virtio/net.c
index f95258c..80c1d81 100644
--- a/virtio/net.c
+++ b/virtio/net.c
@@ -333,7 +333,7 @@ static bool virtio_net__tap_init(struct net_dev *ndev)
 			goto fail;
 	} else if (!skipconf) {
 		memset(&ifr, 0, sizeof(ifr));
-		strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
+		strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ifr.ifr_name));
 		sin.sin_addr.s_addr = inet_addr(params->host_ip);
 		memcpy(&(ifr.ifr_addr), &sin, sizeof(ifr.ifr_addr));
 		ifr.ifr_addr.sa_family = AF_INET;
@@ -345,7 +345,7 @@ static bool virtio_net__tap_init(struct net_dev *ndev)
 
 	if (!skipconf) {
 		memset(&ifr, 0, sizeof(ifr));
-		strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ndev->tap_name));
+		strncpy(ifr.ifr_name, ndev->tap_name, sizeof(ifr.ifr_name));
 		ioctl(sock, SIOCGIFFLAGS, &ifr);
 		ifr.ifr_flags |= IFF_UP | IFF_RUNNING;
 		if (ioctl(sock, SIOCSIFFLAGS, &ifr) < 0)

From patchwork Thu Jan 17 12:40:14 2019
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 8bit
X-Patchwork-Submitter: Anisse Astier <aastier@freebox.fr>
X-Patchwork-Id: 10768015
Return-Path: <kvm-owner@kernel.org>
Received: from mail.wl.linuxfoundation.org (pdx-wl-mail.web.codeaurora.org
 [172.30.200.125])
	by pdx-korg-patchwork-2.web.codeaurora.org (Postfix) with ESMTP id 909176C2
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:38 +0000 (UTC)
Received: from mail.wl.linuxfoundation.org (localhost [127.0.0.1])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 80F6F2F195
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:38 +0000 (UTC)
Received: by mail.wl.linuxfoundation.org (Postfix, from userid 486)
	id 758232F154; Thu, 17 Jan 2019 12:40:38 +0000 (UTC)
X-Spam-Checker-Version: SpamAssassin 3.3.1 (2010-03-16) on
	pdx-wl-mail.web.codeaurora.org
X-Spam-Level: 
X-Spam-Status: No, score=-7.9 required=2.0 tests=BAYES_00,MAILING_LIST_MULTI,
	RCVD_IN_DNSWL_HI autolearn=ham version=3.3.1
Received: from vger.kernel.org (vger.kernel.org [209.132.180.67])
	by mail.wl.linuxfoundation.org (Postfix) with ESMTP id 016FD2F195
	for <patchwork-kvm@patchwork.kernel.org>;
 Thu, 17 Jan 2019 12:40:38 +0000 (UTC)
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
        id S1727835AbfAQMkg (ORCPT
        <rfc822;patchwork-kvm@patchwork.kernel.org>);
        Thu, 17 Jan 2019 07:40:36 -0500
Received: from smtp3-g21.free.fr ([212.27.42.3]:63905 "EHLO smtp3-g21.free.fr"
        rhost-flags-OK-OK-OK-OK) by vger.kernel.org with ESMTP
        id S1727813AbfAQMke (ORCPT <rfc822;kvm@vger.kernel.org>);
        Thu, 17 Jan 2019 07:40:34 -0500
Received: from anisse-station.iliad.local (unknown [213.36.7.13])
        by smtp3-g21.free.fr (Postfix) with ESMTPS id 007C313F8F5;
        Thu, 17 Jan 2019 13:40:28 +0100 (CET)
From: Anisse Astier <aastier@freebox.fr>
To: kvm@vger.kernel.org
Cc: Anisse Astier <aastier@freebox.fr>
Subject: [PATCH kvmtool 3/3] kvmtool: 9p: fix overapping snprintf
Date: Thu, 17 Jan 2019 13:40:14 +0100
Message-Id: <20190117124014.18308-4-aastier@freebox.fr>
X-Mailer: git-send-email 2.19.1
In-Reply-To: <20190117124014.18308-1-aastier@freebox.fr>
References: <20190117124014.18308-1-aastier@freebox.fr>
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Sender: kvm-owner@vger.kernel.org
Precedence: bulk
List-ID: <kvm.vger.kernel.org>
X-Mailing-List: kvm@vger.kernel.org
X-Virus-Scanned: ClamAV using ClamSMTP

GCC 8.2 gives this warning:

virtio/9p.c: In function ‘virtio_p9_create’:
virtio/9p.c:335:21: error: passing argument 1 to restrict-qualified parameter aliases with argument 4 [-Werror=restrict]
  ret = snprintf(dfid->path, size, "%s/%s", dfid->path, name);
                 ~~~~^~~~~~                 ~~~~~~~~~~

Fix it by allocating a temporary string with dfid->path content instead
of overwriting it in-place, which is limited in glibc snprintf with the
__restrict qualifier.

Signed-off-by: Anisse Astier <aastier@freebox.fr>
---
 virtio/9p.c | 9 ++++++++-
 1 file changed, 8 insertions(+), 1 deletion(-)

diff --git a/virtio/9p.c b/virtio/9p.c
index 69fdc4b..6ae1999 100644
--- a/virtio/9p.c
+++ b/virtio/9p.c
@@ -322,6 +322,7 @@ static void virtio_p9_create(struct p9_dev *p9dev,
 	struct p9_qid qid;
 	struct p9_fid *dfid;
 	char full_path[PATH_MAX];
+	char *tmp_path;
 	u32 dfid_val, flags, mode, gid;
 
 	virtio_p9_pdu_readf(pdu, "dsddd", &dfid_val,
@@ -332,7 +333,13 @@ static void virtio_p9_create(struct p9_dev *p9dev,
 		goto err_out;
 
 	size = sizeof(dfid->abs_path) - (dfid->path - dfid->abs_path);
-	ret = snprintf(dfid->path, size, "%s/%s", dfid->path, name);
+	tmp_path = strdup(dfid->path);
+	if (!tmp_path) {
+		errno = ENOMEM;
+		goto err_out;
+	}
+	ret = snprintf(dfid->path, size, "%s/%s", tmp_path, name);
+	free(tmp_path);
 	if (ret >= (int)size) {
 		errno = ENAMETOOLONG;
 		if (size > 0)
